src/mono/wasm/runtime/memory.ts

```diff
@@ -45,44 +45,52 @@ export function _release_temp_frame(): void {
 type _MemOffset = number | VoidPtr | NativePointer | ManagedPointer;
 type _NumberOrPointer = number | VoidPtr | NativePointer | ManagedPointer;
 
-function is_int_in_range(value: Number, min: Number, max: Number) {
-    mono_assert(typeof value === "number", () => `Value is not integer but ${typeof value}`);
-    mono_assert(Number.isInteger(value), "Value is not integer but float");
+function assert_int_in_range(value: Number, min: Number, max: Number) {
+    mono_assert(Number.isSafeInteger(value), () => `Value is not an integer: ${value} (${typeof (value)})`);
     mono_assert(value >= min && value <= max, () => `Overflow: value ${value} is out of ${min} ${max} range`);
 }
 
 export function setB32(offset: _MemOffset, value: number | boolean): void {
-    mono_assert(typeof value === "boolean", () => `Value is not boolean but ${typeof value}`);
-    Module.HEAP32[<any>offset >>> 2] = <any>!!value;
+    const boolValue = !!value;
+    mono_assert(value === boolValue, () => `Value is not boolean but ${typeof value}`);
+    Module.HEAP32[<any>offset >>> 2] = boolValue ? 1 : 0;
 }
 
 export function setU8(offset: _MemOffset, value: number): void {
-    is_int_in_range(value, 0, 0xFF);
+    assert_int_in_range(value, 0, 0xFF);
     Module.HEAPU8[<any>offset] = value;
 }
 
 export function setU16(offset: _MemOffset, value: number): void {
-    is_int_in_range(value, 0, 0xFFFF);
+    assert_int_in_range(value, 0, 0xFFFF);
     Module.HEAPU16[<any>offset >>> 1] = value;
 }
 
+export function setU32_unsafe(offset: _MemOffset, value: _NumberOrPointer): void {
+    Module.HEAPU32[<any>offset >>> 2] = <number><any>value;
+}
+
 export function setU32(offset: _MemOffset, value: _NumberOrPointer): void {
-    is_int_in_range(<any>value, 0, 0xFFFF_FFFF);
+    assert_int_in_range(<any>value, 0, 0xFFFF_FFFF);
     Module.HEAPU32[<any>offset >>> 2] = <number><any>value;
 }
 
 export function setI8(offset: _MemOffset, value: number): void {
-    is_int_in_range(value, -0x80, 0x7F);
+    assert_int_in_range(value, -0x80, 0x7F);
     Module.HEAP8[<any>offset] = value;
 }
 
 export function setI16(offset: _MemOffset, value: number): void {
-    is_int_in_range(value, -0x8000, 0x7FFF);
+    assert_int_in_range(value, -0x8000, 0x7FFF);
     Module.HEAP16[<any>offset >>> 1] = value;
 }
 
+export function setI32_unsafe(offset: _MemOffset, value: number): void {
+    Module.HEAP32[<any>offset >>> 2] = value;
```

Comment:
This will trim anything above 2^31. Instead we need to decide if pointers are signed or unsigned in JS.